### ISSUE TITLE
How can I add and resolve a default file image and avoid the error "path argument must be of type string. Received type object"?

### ISSUE THREAD

**Issue by swquinn** [gatsbyjs/gatsby#26730]
## Summary
I've been trying to customize the `MdxFrontmatter` type to include a `coverImage` field that by default resolves to a `_cover.png` if the `coverImage` front matter content isn't specified (setting a `coverImage` should override it).

For instance, if I have the files `post.mdx` and `_cover.png` in the directory `src/contents/posts/hello-world`, I want the following:

```mdx
// file: src/content/posts/hello-world/post.mdx
---
title: 'Hello world!'
---

# Hello world!
Hi! :)
```

To implicitly resolve the  `coverImage` front matter field to the file represented by `src/contents/posts/hello-world/_cover.png`.

Similarly, if instead of the file `_cover.png` I had the file `foo.png`, and if I explicitly define the `coverImage` front matter as follows:

```mdx
// file: src/content/posts/hello-world/post.mdx
---
title: 'Hello world!'
coverImage: foo.png
---

# Hello world!
Hi! :)
```

I would expect to resolve the `coverImage` to `src/contents/posts/hello-world/foo.png`.

Finally, if I have no image accompanying the `post.mdx` file, I would expect the `coverImage` to resolve to a `null` value.

After making the changes to do the above, whenever I run a query against the `MdxFrontmatter` in the GraphiQL query interface, and include the `coverImage` field, like so:

```gql
{
  allPosts: allMdx {
    edges {
      node {
        id
        frontmatter {
          title
          coverImage {
            id
          } 
        } 
      }
    }
  }
}
```

I get the following error:

```jsonc
{
  "errors": [
    {
      "message": "The \"path\" argument must be of type string. Received type object",
      "locations": [
        {
          "line": 8,
          "column": 11
        }
      ],
      "path": [
        "allPosts",
        "edges",
        18,
        "node",
        "frontmatter",
        "coverImage"
      ],
      "stack": [
        "TypeError [ERR_INVALID_ARG_TYPE]: The \"path\" argument must be of type string. Received type object",
        "    at assertPath (path.js:39:11)",
        "    at Object.resolve (path.js:1088:7)",
        "    at resolve (/Users/seanquinn/Projects/Extesla/lorecall-blog-gatsby/node_modules/gatsby/src/schema/resolvers.ts:344:20)",
        "    at resolveValue (/Users/seanquinn/Projects/Extesla/lorecall-blog-gatsby/node_modules/gatsby/src/schema/resolvers.ts:365:7)",
        "    at map (/Users/seanquinn/Projects/Extesla/lorecall-blog-gatsby/node_modules/gatsby/src/schema/resolvers.ts:364:22)",
        "    at Array.map (<anonymous>)",
        "    at resolveValue (/Users/seanquinn/Projects/Extesla/lorecall-blog-gatsby/node_modules/gatsby/src/schema/resolvers.ts:364:13)",
        "    at fileByPathResolver (/Users/seanquinn/Projects/Extesla/lorecall-blog-gatsby/node_modules/gatsby/src/schema/resolvers.ts:355:12)",
        "    at process._tickCallback (internal/process/next_tick.js:68:7)"
      ]
    },
    // ...
}
```

I'm not sure how to resolve this error: **The \"path\" argument must be of type string. Received type object**

## Relevant information

I've read up pretty thoroughly on [schema customization](https://www.gatsbyjs.com/docs/schema-customization/), looked at source code for various plugins and transformers, and in general believe I have something that is close to working.

I might have misunderstood it, but as far as I could tell the `resolve` function for a field on the Gatsby Type Builder does the same thing that one would expect to do in the `createResolvers`. Below, you can see how I am creating the `coverImage` field on the `MdxFrontmatter` Node. The query with the filter appears to work as I would expect, and I get no errors during `$ gatsby build` or `$ gatsby develop`.

```javascript
  createTypes([
    schema.buildObjectType({
      name: `MdxFrontmatter`,
      fields: {
        coverImage: {
          type: `File`,
          extensions: { fileByRelativePath: {} },
          resolve: async (source, args, context, info) => {
            const fieldValue = source && source[info.fieldName] || '_cover.png'
            const sourceFileNode = context.nodeModel.findRootNodeAncestor(
              source,
              node => node.internal && node.internal.type === `File`
            )
            const coverImageRelativePath = path.join(sourceFileNode.relativeDirectory, fieldValue)
            const fileNode = await context.nodeModel.runQuery({
              type: `File`,
              query: {
                filter: {
                  relativePath: {
                    eq: coverImageRelativePath,
                  },
                },
              },
            })
            return fileNode
          }
        },
      },
      interfaces: [`Node`],
    })
  ])
```

If I `console.log` the result of the run query which gets assigned to `fileNode`, I see the file descriptor information for the files that match either the default `coverImage` or the specified `coverImage` in the front matter of a post. Anything that doesn't match yields a `null` value (which I would expect.

I've tried adding a `firstOnly: true` to the query, but that doesn't seem to change anything.

### Environment

```
  System:
    OS: macOS 10.15.6
    CPU: (8) x64 Intel(R) Core(TM) i7-4870HQ CPU @ 2.50GHz
    Shell: 3.2.57 - /bin/bash
  Binaries:
    Node: 10.13.0 - ~/.nvm/versions/node/v10.13.0/bin/node
    Yarn: 1.19.2 - /usr/local/bin/yarn
    npm: 6.4.1 - ~/.nvm/versions/node/v10.13.0/bin/npm
  Languages:
    Python: 2.7.17 - /usr/local/bin/python
  Browsers:
    Chrome: 85.0.4183.83
    Firefox: 80.0
    Safari: 13.1.2
  npmPackages:
    gatsby: ^2.24.51 => 2.24.53 
    gatsby-cli: ~2.12.89 => 2.12.91 
    gatsby-image: ~2.4.16 => 2.4.16 
    gatsby-plugin-google-analytics: ~2.3.13 => 2.3.13 
    gatsby-plugin-mdx: ^1.2.37 => 1.2.37 
    gatsby-plugin-sharp: ~2.6.31 => 2.6.31 
    gatsby-remark-images: ^3.3.28 => 3.3.28 
    gatsby-source-filesystem: ~2.3.26 => 2.3.27 
    gatsby-transformer-sharp: ^2.5.14 => 2.5.14 
  npmGlobalPackages:
    gatsby-cli: 2.9.0
```

### File contents (if changed)

`gatsby-config.js`:

```
require('dotenv').config({
  path: `.env`,
})

module.exports = {
  siteMetadata: {
    title: `Gatsby`,
    description: `Kick off your next, great Gatsby project with this default starter. This barebones starter ships with the main Gatsby configuration files you might need.`,
    author: `@gatsbyjs`,
  },
  plugins: [
    {
      resolve: `gatsby-plugin-mdx`,
      options: {
        extensions: ['.mdx', '.md'],
        gatsbyRemarkPlugins: [
          {
            resolve: 'gatsby-remark-images',
            options: {
              maxWidth: 1200,
            },
          },
        ],
      },
    },
    `gatsby-transformer-sharp`,
    `gatsby-plugin-sharp`,
    'gatsby-remark-images',
    {
      resolve: `gatsby-source-filesystem`,
      options: {
        name: `posts`,
        path: `${__dirname}/src/contents/posts`,
      },
    },
    {
      resolve: `gatsby-source-filesystem`,
      options: {
        name: `images`,
        path: `${__dirname}/src/images`,
      },
    },
  ],
}
```

`package.json`:

> Using `yarn`

```json
{
  "name": "gatsby-blog-test",
  "version": "1.0.0",
  "description": "",
  "repository": "",
  "license": "UNLICENSED",
  "scripts": {
    "build": "node_modules/.bin/gatsby build",
    "develop": "node_modules/.bin/gatsby develop",
    "gatsby": "node_modules/.bin/gatsby",
    "eslint": "node_modules/.bin/eslint src",
    "start": "yarn develop",
    "test": ""
  },
  "dependencies": {
    "@babel/core": "^7.10.5",
    "@mdx-js/mdx": "^1.6.16",
    "@mdx-js/react": "^1.6.16",
    "@mdx-js/tag": "^0.20.3",
    "gatsby": "^2.24.51",
    "gatsby-image": "~2.4.16",
    "gatsby-plugin-mdx": "^1.2.37",
    "gatsby-plugin-sharp": "~2.6.31",
    "gatsby-remark-images": "^3.3.28",
    "gatsby-source-filesystem": "~2.3.26",
    "gatsby-transformer-sharp": "^2.5.14",
    "react": "~16.13.1",
    "react-dom": "~16.13.1",
    "redux": "~4.0.5",
    "remark": "^12.0.1",
    "remark-html": "^12.0.0"
  },
  "devDependencies": {
    "@babel/eslint-parser": "^7.11.0",
    "@babel/eslint-plugin": "^7.11.0",
    "eslint": "^7.7.0",
    "eslint-config-standard": "^14.0.0",
    "eslint-plugin-import": "^2.22.0",
    "eslint-plugin-node": "^11.1.0",
    "eslint-plugin-promise": "^4.2.0",
    "eslint-plugin-react": "^7.20.0",
    "eslint-plugin-standard": "^4.0.0",
    "gatsby-cli": "~2.12.89",
    "webpack": "^4.44.0"
  }
}
```

`gatsby-node.js`:

> Modified to add the default `coverImage` field to the `MdxFrontmatter` Node.

```javascript
const path = require('path')
const { createFilePath } = require('gatsby-source-filesystem')

exports.createSchemaCustomization = ({ actions, schema }) => {
  const { createTypes, createFieldExtension } = actions
  createTypes([
    schema.buildObjectType({
      name: `MdxFrontmatter`,
      fields: {
        coverImage: {
          type: `File`,
          extensions: { fileByRelativePath: {} },
          resolve: async (source, args, context, info) => {
            const fieldValue = source && source[info.fieldName] || '_cover.png'
            const sourceFileNode = context.nodeModel.findRootNodeAncestor(
              source,
              node => node.internal && node.internal.type === `File`
            )
            const coverImageRelativePath = path.join(sourceFileNode.relativeDirectory, fieldValue)
            const fileNode = await context.nodeModel.runQuery({
              type: `File`,
              query: {
                filter: {
                  relativePath: {
                    eq: coverImageRelativePath,
                  },
                },
              },
            })
            return fileNode
          }
        },
      },
      interfaces: [`Node`],
    })
  ])
}
```

**Comment by vladar** [comment]
Thank you for opening this, @swquinn 

The error you see is a result of having both - custom resolver **and** `extensions: { fileByRelativePath: {} }`. This extension provides it's own resolver on top of yours which expects your resolver to return `string` with a relative path to your file. But your custom resolver returns object vs string.

So to fix this you have two options:

Option1. Don't use the extension:

```diff
  createTypes([
    schema.buildObjectType({
      name: `MdxFrontmatter`,
      fields: {
        coverImage: {
          type: `File`,
-         extensions: { fileByRelativePath: {} },
          resolve: async (source, args, context, info) => {
            const fieldValue = source && source[info.fieldName] || '_cover.png'
            const sourceFileNode = context.nodeModel.findRootNodeAncestor(
              source,
              node => node.internal && node.internal.type === `File`
            )
            const coverImageRelativePath = path.join(sourceFileNode.relativeDirectory, fieldValue)
            const fileNode = await context.nodeModel.runQuery({
              type: `File`,
              query: {
                filter: {
                  relativePath: {
                    eq: coverImageRelativePath,
                  },
                },
              },
+             firstOnly: true,
            })
            return fileNode
          }
        },
      },
      interfaces: [`Node`],
    })
  ])
```

Option 2. Use the extension and adjust your custom resolver accordingly:

```diff
  createTypes([
    schema.buildObjectType({
      name: `MdxFrontmatter`,
      fields: {
        coverImage: {
          type: `File`,
          extensions: { fileByRelativePath: {} },
          resolve: async (source, args, context, info) => {
-          // old code
+           return source && source[info.fieldName] || '_cover.png'
          }
        },
      },
      interfaces: [`Node`],
    })
  ])
```

Basically the extension does most of what you're doing in your resolver.

We're marking this issue as answered and closing it for now but please feel free to reopen this and comment if you would like to continue this discussion. We hope we managed to help and thank you for using Gatsby! 💜
